topi/include/topi/detail/extern.h

```diff
@@ -0,0 +1,93 @@
+/*!
+*  Copyright (c) 2017 by Contributors
+* \file detail/extern.h
+* \brief Helpers for using external functions
+*/
+#ifndef TOPI_DETAIL_EXTERN_H_
+#define TOPI_DETAIL_EXTERN_H_
+
+#include <vector>
+#include <string>
+
+#include "tvm/tvm.h"
+
+namespace topi {
+using namespace tvm;
+
+/*! \brief Construct a buffer to pass to an external function */
+Buffer DeclExternBuffer(Array<Expr> shape,
+  Type dtype,
+  std::string name) {
+  auto data = var(name, Handle());
+  auto elem_offset = Expr();
+  return BufferNode::make(data, dtype, shape, Array<Expr>(), elem_offset, name, "",
+    -1, 0);
+}
+
+/*! \brief A function wrapping the invocation of an external function */
+using FExtern = std::function<Expr(Array<Buffer>, Array<Buffer>)>;
+
+/*! \brief Create tensors representing the result of invoking an external function */
+Array<Tensor> make_extern(const Array<Array<Expr>>& out_shapes,
+                          const std::vector<Type>& out_types,
+                          const Array<Tensor>& inputs,
+                          FExtern fextern,
+                          std::string name,
+                          std::string tag) {
+  CHECK_EQ(out_shapes.size(), out_types.size())
+    << "make_extern: out_shapes and out_types must have equal size";
+
+  Array<Buffer> input_placeholders;
+  for (auto t : inputs) {
+    input_placeholders.push_back(DeclExternBuffer(t->shape, t->dtype, t->op->name));
+  }
+  Array<Buffer> output_placeholders;
+  for (size_t i = 0; i < out_shapes.size(); ++i) {
+    output_placeholders.push_back(DeclExternBuffer(out_shapes[i], out_types[i], name));
+  }
+
+  auto body = fextern(input_placeholders, output_placeholders);
+  auto body_stmt = tvm::ir::Evaluate::make(body);
+
+  auto op = ExternOpNode::make(
+    name, tag, inputs, input_placeholders, output_placeholders, body_stmt);
+
+  Array<Tensor> outputs;
+  for (size_t i = 0; i < output_placeholders.size(); ++i) {
+    outputs.push_back(op.output(i));
+  }
+  return outputs;
+}
+
+/*! \brief Pack a buffer object to be used as an argument to a PackedFunc */
+Expr pack_buffer(Buffer buf) {
```

Comment:
this belongs to a detail namespace. Add more comments to argument and return value.

This functions is used to create a DLTensor structure on heap to be able to pass a symbolic buffer as arguments to TVM PackedFunc